connectors/sources/postgresql.py

```diff
@@ -0,0 +1,81 @@
+#
+"""SQLAlchemy source module is responsible to fetch documents from PostgreSQL."""
+import ssl
+from urllib.parse import quote
+
+from sqlalchemy.ext.asyncio import create_async_engine
+
+from connectors.logger import logger
+from connectors.sources.generic_database import GenericBaseDataSource
+
+# Below schemas are system schemas and the tables of the systems schema's will not get indexed
+SYSTEM_SCHEMA = ["pg_toast", "pg_catalog", "information_schema"]
+
+
+class PostgreSQLDataSource(GenericBaseDataSource):
+    """Class to fetch documents from Postgresql Server"""
+
+    def __init__(self, connector):
+        """Setup connection to the PostgreSQL database-server configured by user
+
+        Args:
+            connector (BYOConnector): Object of the BYOConnector class
+        """
+        super().__init__(connector=connector)
+        self.connection_string = f"postgresql+asyncpg://{self.user}:{quote(self.password)}@{self.host}:{self.port}/{self.database}"
+        self.queries = {
+            "PING": "SELECT 1+1",
+            "ALL_TABLE": "SELECT table_name FROM information_schema.tables WHERE table_catalog = '{database}' and table_schema = '{schema}'",
+            "TABLE_PRIMARY_KEY": "SELECT c.column_name FROM information_schema.table_constraints tc JOIN information_schema.constraint_column_usage AS ccu USING (constraint_schema, constraint_name) JOIN information_schema.columns AS c ON c.table_schema = tc.constraint_schema AND tc.table_name = c.table_name AND ccu.column_name = c.column_name WHERE constraint_type = 'PRIMARY KEY' and tc.table_name = '{table}' and tc.constraint_schema = '{schema}'",
+            "TABLE_DATA": 'SELECT * FROM {schema}."{table}"',
+            "TABLE_LAST_UPDATE_TIME": 'SELECT MAX(pg_xact_commit_timestamp(xmin)) FROM {schema}."{table}"',
+            "TABLE_DATA_COUNT": 'SELECT COUNT(*) FROM {schema}."{table}"',
+            "ALL_SCHEMAS": "SELECT schema_name FROM information_schema.schemata",
+        }
+        self.engine = None
+
+    async def ping(self):
+        """Verify the connection with the PostgreSQL database-server configured by user"""
+        logger.info("Validating the Connector Configuration...")
+        try:
+            self._validate_configuration()
+            self.engine = create_async_engine(
+                self.connection_string,
+                connect_args=self.get_connect_args() if not self.ssl_disabled else {},
+            )
+            await anext(self.execute_query(query_name="PING", engine=self.engine))
+            logger.info("Successfully connected to PostgreSQL.")
+        except Exception:
+            raise Exception(f"Can't connect to Postgresql Server on {self.host}")
+
+    def get_connect_args(self):
+        """Convert string to pem format and create a SSL context
+
+        Returns:
+            dictionary: Connection arguments
+        """
+        self.ssl_ca = self.ssl_ca.replace(" ", "\n")
```

Comment:
Why do we do some specific extraction here? Shouldn't that be done in the generic connector? it looks like you are removing the `--BEGIN` etc? should be a function with an explicit name, like `parse_pem` that returns the key etc

Reply:
> Why do we do some specific extraction here?

This method is specific to PostgreSQL dialect only, so we should not move it to generic connector file.

> Shouldn't that be done in the generic connector? it looks like you are removing the --BEGIN etc?

Also, we're not removing `---BEGIN--` and `---END---` from the string, we're making it like `pem` format to parse it into `cadata` - https://github.com/elastic/connectors-python/blob/c65bd1ea4b538cbc395f47d654aa2813cf2ec826/connectors/sources/postgresql.py#L65

> should be a function with an explicit name, like parse_pem that returns the key etc

The function takes care of creating a connection argument for engine that's why we named this way.

Reply:
can you create a separate 

```
pem_format = self.get_pem_format(self.ssl_ca)
```

with a unit test that passes a SSL cert and returns the pem format.
This will clarify what we are doing here.

thx

Reply:
Done